common/src/main/java/io/netty/util/Recycler.java

```diff
@@ -236,41 +237,93 @@ public void recycle(Object object) {
+                    head = link.next;
+                }
+            }
+
+            void reclaimSpace(int space) {
+                assert space >= 0;
+                availableSharedCapacity.addAndGet(space);
+            }
+
+            boolean reserveSpace(int space) {
+                return reserveSpace(availableSharedCapacity, space);
+            }
+
+            static boolean reserveSpace(AtomicInteger availableSharedCapacity, int space) {
+                assert space >= 0;
+                for (;;) {
+                    int available = availableSharedCapacity.get();
+                    if (available < space) {
+                        return false;
+                    }
+                    if (availableSharedCapacity.compareAndSet(available, available - space)) {
+                        return true;
+                    }
+                }
+            }
+        }
+
         // chain of data items
-        private Link head, tail;
+        private final Head head;
+        private Link tail;
         // pointer to another queue of delayed items for the same stack
         private WeakOrderQueue next;
         private final WeakReference<Thread> owner;
         private final int id = ID_GENERATOR.getAndIncrement();
-        private final AtomicInteger availableSharedCapacity;
 
         private WeakOrderQueue() {
             owner = null;
-            availableSharedCapacity = null;
+            head = new Head(null);
         }
 
         private WeakOrderQueue(Stack<?> stack, Thread thread) {
-            head = tail = new Link();
-            owner = new WeakReference<Thread>(thread);
+            tail = new Link();
 
             // Its important that we not store the Stack itself in the WeakOrderQueue as the Stack also is used in
             // the WeakHashMap as key. So just store the enclosed AtomicInteger which should allow to have the
             // Stack itself GCed.
-            availableSharedCapacity = stack.availableSharedCapacity;
+            head = new Head(stack.availableSharedCapacity);
+            head.link = tail;
+            owner = new WeakReference<Thread>(thread);
         }
 
         static WeakOrderQueue newQueue(Stack<?> stack, Thread thread) {
-            WeakOrderQueue queue = new WeakOrderQueue(stack, thread);
+            final WeakOrderQueue queue = new WeakOrderQueue(stack, thread);
             // Done outside of the constructor to ensure WeakOrderQueue.this does not escape the constructor and so
             // may be accessed while its still constructed.
             stack.setHead(queue);
+
+            // We need to reclaim all space that was reserved by this WeakOrderQueue so we not run out of space in
+            // the stack. This is needed as we not have a good life-time control over the queue as it is used in a
+            // WeakHashMap which will drop it at any time.
+            final Head head = queue.head;
+            ObjectCleaner.register(queue, head);
```

Comment:
I wonder if it makes sense to be able to create an ObjectCleaner instance ... so we can partition its usage and reduce contention on the static state within (e.g. the `Set`, ...). any thoughts? This would give us flexibility to partition it depending upon the use case (e.g. a static instance for all `Recycler` objects, another static instance for all `FastThreadLocal` objects, individual `Recycler` objects, etc...).

Reply:
It seems like ConcurrentHashMap should do a good job of avoiding contention. If it becomes a problem then we can adjust the load factor. The bigger thing to me is only having one cleaner thread; I'd hope to keep it that way.